Soarian FHIR R4 changes for Overview, R4 Overview and Patient resource (#677)
cerner/fhir.cerner.com: content/soarian/r4.md

---
title: R4
layout: api
---

# R4 Overview

* TOC
{:toc}

Cerner’s Soarian Clinicals<sup>®</sup> implementation currently supports both the R4 First Normative Content (4.0.1) version and DSTU 2 Final (1.0.2) version of the Health Level Seven (HL7)<sup>®</sup> International Fast Healthcare Interoperability Resources (FHIR)<sup>®</sup> standard. Cerner’s implementation of the R4 version is ongoing and new resources and actions will be added over time.

> Existing DSTU 2 patient-facing applications will eventually need to be migrated to our R4 implementation. For this reason, Cerner strongly encourages development against R4 whenever possible.

Reviewer: ```suggestion
Existing DSTU 2 patient-facing applications will eventually need to be migrated to the Soarian Clinicals<sup>®</sup> FHIR<sup>®</sup> R4 implementation. For this reason, Cerner strongly encourages development against R4 whenever possible.
```

Author: Done with the changes

Reviewer: Looks good to me.

> ## Schema

Reviewer 2: are you sure you want to put the whole capability statement here? It's only going to get bigger

Reviewer 3: Clip the contents for better readability

Author: I discussed with Sudhir and reduced the  capability statement content

> All API access is over HTTPS. See [Service Root URL](#service-root-url) for more information on the URL format. All data is sent and received as JSON.

Author: Done with the changes

            },
            {
              "name": "_count",
              "definition": "http://hl7.org/fhir/search.html#count",
              "type": "integer",
              "documentation": "Number of resources should be returned in single page"
            },
            {
              "name": "_revinclude",
              "definition": "http://hl7.org/fhir/search.html#revinclude",
              "type": "reference",
              "documentation": "Includes Provenance resource in response. It may only be provided if at least one other parameter is provided."
            }
          ],
          "searchRevInclude": [
            "Provenance:target"
          ]
        }
      ]
    }
  ]
}

</pre>

Blank fields are omitted.

All timestamps are returned in [FHIR<sup>®</sup> standard date/dateTime formats](http://www.hl7.org/implement/standards/fhir/r4/datatypes.html#date).

### Media Types

> Cerner supports the R4 FHIR<sup>®</sup> standard defined media type for JSON content:

Author: Done

    application/fhir+json

> Cerner recommends that you explicitly request this accept type using the `Accept` header.

Author: Done

## Service Root URL

URLs for the FHIR<sup>®</sup> server vary according to the tenant (data source or client) being accessed, as well as other factors. If the application is a Substitutable Medical Applications and Reusable Technologies (SMART)<sup>®</sup> application, the [service root url](http://hl7.org/fhir/r4/http.html#general) is provided at launch time. For standalone applications, the URL can be requested or configured when the application is set up to run against a specific tenant. SMART<sup>®</sup> applications make FHIR<sup>®</sup> calls against URLs of the following format:

`:serviceRootURL/:resource[?:parameters]`

### Secure Sandbox

> The secure sandbox instance can be used for testing an application with [authorization](#authorization). The service root URL for this instance is different if the patient or a patient’s proxy is logging in.

Reviewer: ```suggestion
You can use the secure sandbox for testing an application with [authorization](#authorization). The service root URL for this instance is different if the patient or a patient’s proxy is logging in.
```

Author: Done with the changes

Reviewer: Looks good to me.

> Non-Patient: `https://fhir-ehr-sc.cerner.com/r4/2f8f5ec1-b7b8-4be5-ae27-e308284dd9c1/:resource[?:parameters]`

Author: Done

> Patient Access: `https://fhir-myrecord-sc.cerner.com/r4/2f8f5ec1-b7b8-4be5-ae27-e308284dd9c1/:resource[?:parameters]`

Author: Done

    https://fhir-myrecord-sc.cerner.com/r4/2f8f5ec1-b7b8-4be5-ae27-e308284dd9c1/Patient/4C05F5F357EE44FC874C0CF4AB249F99

In another context, the ID  **Patient/4C05F5F357EE44FC874C0CF4AB249F99** may identify another person entirely. In the following example, a different resource may be returned because the context (service root URL) has changed.

    https://fhir-myrecord-sc.cerner.com/r4/ec2458f2-1e24-41c8-b71b-0e701af7583d/Patient/4C05F5F357EE44FC874C0CF4AB249F99

Similarly, when considering an identifier, you must consider it only in its full context. Even though some identifiers may exist across multiple systems (for example, MRN), they may not refer to the same resource.

    https://fhir-myrecord-sc.cerner.com/r4/2f8f5ec1-b7b8-4be5-ae27-e308284dd9c1/Patient?identifier=urn:oid:1.1.1.1.1.1|1001006

For example, when using the above MRN in a different system, the same Patient resource may not be returned in the response bundle:

    https://fhir-myrecord-sc.cerner.com/r4/ec2458f2-1e24-41c8-b71b-0e701af7583d/Patient?identifier=urn:oid:1.1.1.1.1.1|1001006

## Client Errors

Several possible types of client errors may occur on API calls that receive request bodies:

> * Failing to send a required query parameter when header processing preference is set to “handling=lenient,” or requesting a media type other than JSON, results in a `400 Bad Request` response.

Reviewer: ```suggestion
* Failing to send a required query parameter when the header processing preference is set to “handling=lenient,” or requesting a media type other than JSON, results in a `400 Bad Request` response.
```

Author: Done with the changes

        HTTP/1.1 404 Not Found
		

* Requesting a media type other than json, application/json, application/fhir+json, application/json+fhir in the HTTP accept header will result in a `406 Not Acceptable` response.

        HTTP/1.1 406 Not Acceptable

        Content-Length: 0

* Posting an unsupported media format will result in a `415 Unsupported Media Type` response.

        HTTP/1.1 415 Unsupported Media Type
		

## Operation Outcomes

An [OperationOutcome](https://www.hl7.org/fhir/r4/operationoutcome.html) resource may be returned to provide additional context to an error. The following tables describe common operation outcomes and their causes.  

 HTTP Status                    | Cause                                                                      | Severity | Code
--------------------------------|----------------------------------------------------------------------------|----------|----------
 400                            | Format of the parameter list is invalid                                    | error    | structure
 400                            | Required parameter missing                                                 | error    | required

>  400<br><br><br><br><br>        |  •  Code value or coding system not supported <br> •  Duplicate parameters  error <br> •  Unsupported interaction  <br> •  Unsupported operation  <br> •  Unsupported parameter                                                        | error    |not-supported                                

Reviewer: Correct bullet/line break

>  400<br><br><br><br><br>        |  •  Code value or coding system not supported <br> •  Duplicate parameters  error <br> •  Unsupported interaction  <br> •  Unsupported operation  <br> •  Unsupported parameter                                                        | error    |not-supported                                

Reviewer: Correct bullets/line breaks

Author: Done with the changes

Reviewer: Looks better, thanks.

>  400<br><br><br><br><br>        |  •  Code value or coding system not supported <br> •  Duplicate parameters  error <br> •  Unsupported interaction  <br> •  Unsupported operation  <br> •  Unsupported parameter                                                        | error    |not-supported                                

Author: As discussed, we are going with the same table

 400                            | Unexpected multiple matches for a given ID                                 | error    | multiple-matches
 400                            | Too many results; narrow search                                            | error    | business-rule
 401                            | Invalid token                                                              | error    | Unknown
 403<br><br>                    |  •  Persona missing from JWT claims <br> •  Unrecognized scope             | error    | Security

>  404<br><br>                    |  •  Patient ID was merged to a target <br> •  No matching results     | error    | not-found

Reviewer: Correct bullets/line breaks

Author: Done with the changes

 

## Handling Required Fields

1. An [OperationOutcome](https://www.hl7.org/fhir/r4/operationoutcome.html) resource may be returned to provide information regarding missing fields that are required by HL7<sup>®</sup> FHIR<sup>®</sup> or the [US Core](https://www.hl7.org/fhir/us/core/) HL7<sup>®</sup> profile. This OperationOutcome resource contains the [DataAbsentReason](http://hl7.org/fhir/r4/extension-data-absent-reason.html).

        {
          "coding": [
            {
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/data-absent-reason",
                  "valueCode": "unknown"
                }
              ]
            }
          ]
        }
    

## Filtered Data for Patient or Proxy Access

> When accessing data with a patient persona, resources with a status of **entered-in-error** are filtered in contrast to system or provider access, where they are returned.

Author: Yes. It is not applicable to Soarian Clinicals. So removed the section

## General Security Filtering

Certain data may not be available to all applications or users. Encounter or organization security, privileges, and preferences may filter data.

## Authorization

> Cerner Ignite APIs for Soarian Clinicals<sup>®</sup> has an endpoint secured with [OAuth 2.0](https://oauth.net/2/) with support for [SMART<sup>®</sup> Applications](http://docs.smarthealthit.org/). Refer to the extensions on the `Conformance.rest.security` element in our server [metadata](http://fhir.cerner.com/soarian/r4/conformance/).

Author: Done with the changes

> Each resource interaction documents the type of acceptable authentication (patient, provider, and/or system). While an interaction may list system authentication, this functionality is currently not available in production.

Reviewer: ```suggestion
Each resource interaction documents the type of acceptable authentication (patient, provider, or system). While an interaction may list system authentication, this functionality is currently not available in production.
```

Author: Done with the changes

> See the [authorization](http://fhir.cerner.com/authorization/) documentation for details on how to authorize using our server.

Reviewer: ```suggestion
See the [authorization](http://fhir.cerner.com/authorization/) documentation for details on how to authorize using the Soarian Clinicals<sup>®</sup> FHIR<sup>®</sup> R4 server.
```

Author: Done with the changes

## Pagination

> The pagination links are included in the Bundle when a resource returns more items than the Bundle page size. It is important to follow these Link header values, rather than constructing your own URLs.

Author: Done

This is particularly relevant when the server is returning the results in a series of pages. It is at the discretion of the search engine of how to handle ongoing updates to the resources while the search is proceeding."

When results are returned in a series of pages, you may see the same resource ID on subsequent pages. Developers need to take this situation into account by filtering by resource ID and only return the latest version of that resource ID. Displaying duplicate resource IDs or any other manipulation may misrepresent the data, thus affecting clinical decision support and patient safety.

## Cross Origin Resource Sharing

The API supports Cross Origin Resource Sharing (CORS) for AJAX requests from any origin. You can read the [CORS W3C Recommendation](http://www.w3.org/TR/cors), or [this intro](http://code.google.com/p/html5security/wiki/CrossOriginRequestSecurity) from the HTML 5 Security Guide.

Here's a sample request sent from the origin `http://example.com`:

<pre class="terminal">

> $ curl -i -H "Origin: http://example.com" -H "Accept: application/json+fhir" https://fhir-myrecord-sc.devcerner.com/r4/2f8f5ec1-b7b8-4be5-ae27-e308284dd9c1/metadata

Author: Done

HTTP/1.1 200 OK
Access-Control-Allow-Origin: http://example.com
Access-Control-Allow-Methods: GET, POST
Access-Control-Max-Age: 0
Access-Control-Allow-Credentials: true
</pre>

This is what a CORS preflight request looks like:

<pre class="terminal">

> $ curl -X OPTIONS -i -H "Origin: http://example.com"  https://fhir-myrecord-sc.devcerner.com/r4/2f8f5ec1-b7b8-4be5-ae27-e308284dd9c1/metadata

Author: Done
